### ISSUE TITLE
defer() pattern tasks

### ISSUE THREAD

**Issue by tonyandrewmeyer** [canonical/operator#1125]
Follow-up actions from the [defer analysis](https://docs.google.com/document/d/1J9Fm2LuMC6kcUCil61Tf1BQO13UbGZEm7WoZVMtZtoQ/edit#heading=h.5zkls1ejvnfk):

* [x] Investigate removing the defer() method from event classes that can’t defer (changing to AttributeError from RuntimeError, so that IDEs and linters easily detect it)
* [x] Remove the ability to defer (currently only Action, SecretRemoved, SecretExpired) from other relevant classes (CollectStatus, maybe some of the teardown events)
* Discourage “if container.can_connect()” guards
  * [x] Reword the API reference docs so that it doesn’t push people this way
  * [x] Check other docs (e.g. the tutorial) to see if improvements can be made
  * [x] ~Maybe some additional reach-out~
* For “I need this collection of things to be ready”: where possible, encourage a common code path and return without deferring (trusting that the last event will take care of everything).
  * [x] Update docs
    * [x] ["Deferring Events: Details and Dilemmas"](https://juju.is/docs/sdk/deferring-events-details-and-dilemmas) ([Discourse](https://discourse.charmhub.io/t/deferring-events-details-and-dilemmas/5930)) - there are some other references to `defer()` on juju.is.docs/sdk, but they are all explaining what `defer()` does, and already have appropriate wording around taking care with its use, and don't seem to need anything additional
    * [x] ~[SDK](https://ops.readthedocs.io/en/latest/#ops.EventBase.defer)~ - this has a good section on taking care when deciding if `defer()` should be used, and two examples of where it's not a good choice. This is already one of the longest method docstrings we have - I don't think it's the right place to go into more detail (at most we could add a link to more documentation)
  * [x] Review key charms from each team and reach out where applicable with suggestions
  * [x] Maybe a Discourse post encouraging best practices
* [x] [Feature request for a Juju retry](https://warthogs.atlassian.net/browse/PF-4707)

**Comment by tonyandrewmeyer** [comment]
> * [ ]  Investigate removing the defer() method from event classes that can’t defer (changing to AttributeError from RuntimeError, so that IDEs and linters easily detect it)
> * [ ]  Remove the ability to defer (currently only Action, SecretRemoved, SecretExpired) from other relevant classes (CollectStatus, maybe some of the teardown events)

Both in https://github.com/canonical/operator/pull/1122 (TBD if we will remove or not).

> * Discourage “if container.can_connect()” guards
>   
>   * [ ]  Reword the API reference docs so that it doesn’t push people this way

https://github.com/canonical/operator/pull/1123

**Comment by tonyandrewmeyer** [comment]
>   * [ ]  Check other docs (e.g. the tutorial) to see if improvements can be made

Tutorial:
 * [ ] https://discourse.charmhub.io/t/make-your-charm-configurable/7401
 * [ ] https://discourse.charmhub.io/t/expose-the-version-of-the-application-behind-your-charm/7406
 * [ ] https://discourse.charmhub.io/t/integrate-your-charm-with-postgresql/7414

Other:
* [ ] https://discourse.charmhub.io/t/storedstate-uses-limitations/5740
* [ ] https://discourse.charmhub.io/t/event-container-pebble-ready/6468

That seems to be all the cases on Discourse, so hopefully that (plus the already done API docs) should be everywhere.

**Comment by tonyandrewmeyer** [comment]
>     * [ ]  ["Deferring Events: Details and Dilemmas"](https://juju.is/docs/sdk/deferring-events-details-and-dilemmas) ([Discourse](https://discourse.charmhub.io/t/deferring-events-details-and-dilemmas/5930)) - there are some other references to `defer()` on juju.is.docs/sdk, but they are all explaining what `defer()` does, and already have appropriate wording around taking care with its use, and don't seem to need anything additional

@benhoyt could you please take a look at [these suggested changes to that doc](https://docs.google.com/document/d/1TAySeElAiNzNRKP1E_xay2ODuab1KqyAGbLhws4dOJM/edit?usp=sharing)? Happy to schedule a call to go over them together if that works best.

**Comment by benhoyt** [comment]
@tonyandrewmeyer Added a couple of comments. I think there's more work to be done on that doc, but I think the improvements you've made are all in the right direction -- thanks!

**Comment by tonyandrewmeyer** [comment]
>   * [ ]  Review key charms from each team and reach out where applicable with suggestions

I've reviewed all the use of `defer()` in the "key charms" (summaries are in the charm spreadsheet), and there's a lot of `can_connect()` + `defer()` that I would suggest is better handled by catching `ConnectionError` but would still be `defer()`. Several charms already use the holistic pattern, and where the others aren't, there are cases where (given the unpredictability of ordering) `defer()` is still required.

In short: I didn't find any charm where there are significant areas for improvement in `defer()` use - outside of a first-class Juju reemit method being added.

**Comment by tonyandrewmeyer** [comment]
>   * [ ]  Maybe a Discourse post encouraging best practices

@benhoyt would you be able to find some time this week to review [this guidance doc](https://docs.google.com/document/d/1T2W7eCgBfKpWct5KbHMz1uYPXl-IpAtwzZr6CsdOmhs/edit?usp=sharing)? Thanks!
